src/freenet/node/PeerMessageQueue.java

```diff
@@ -876,22 +876,20 @@ public synchronized MessageItem grabQueuedMessageItem(int minPriority) {
 			if(ret != null) return ret;
 		}
 		
-		// Include bulk or realtime, whichever is more urgent.
-		
-		boolean tryRealtimeFirst = this.fastWeakRandom.nextInt(10) > 0;
-		
-		// If one is empty, try the other.
-		// Otherwise try whichever is more urgent, favouring realtime if there is a draw.
-		// Realtime is supposed to be bursty.
 		
+
+		// Include bulk or realtime, whichever is more urgent.
+		boolean tryRealtimeFirst;
 		if(queuesByPriority[DMT.PRIORITY_REALTIME_DATA].isEmpty()) {
 			tryRealtimeFirst = false;
 		} else if(queuesByPriority[DMT.PRIORITY_BULK_DATA].isEmpty()) {
 			tryRealtimeFirst = true;
-		} else if(queuesByPriority[DMT.PRIORITY_BULK_DATA].getNextUrgentTime(Long.MAX_VALUE, 0) >= queuesByPriority[DMT.PRIORITY_REALTIME_DATA].getNextUrgentTime(Long.MAX_VALUE, 0)) {
+		} else if(queuesByPriority[DMT.PRIORITY_BULK_DATA].getNextUrgentTime(Long.MAX_VALUE, 0)
+				>= queuesByPriority[DMT.PRIORITY_REALTIME_DATA].getNextUrgentTime(Long.MAX_VALUE, 0)) {
 			tryRealtimeFirst = true;
 		} else {
-			tryRealtimeFirst = false;
+			// 10% chance to use bulk in case of a draw to avoid starving the bulk queue.
+			tryRealtimeFirst = this.fastWeakRandom.nextInt(10) > 0;
 		}
```

Comment:
This resolves a dumb mistake: the exhaustive if-else block always replaced the random value.